patterns/interpreter.md

```diff
@@ -0,0 +1,271 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve it,
+then the problem instances might be expressed in a simple
+language and inerpreter object could solve it by
+interprting the sentences written in this simple languge.
+Basically, for any kind of problems we define a domain
+language, then define a grammar for this language and
+design interpreter solving problem instances.
+
+## Motivation
+
+Imagine thta our work is translating simple mathemtaical expressions into
+[assembly language](https://en.wikipedia.org/wiki/Assembly_language)
+(more simple and low level programming language).
+For simplicity, our expressions consists of ten digits `0`,...,`9`,
+four operations `+, -, /, *` and a pair of prenthesis `(, )`.
+For example, experssion `2 + 4` could be translated into
+
+```ignore
+mov eax, 2
+mov ebx, 4
+add eax, ebx
+```
+
+Our goal is to automate translating into assembly instructions
+using the Interpreter design pattern. In other words, we want simply
+provide the Interpreter with an expressions and get assembly
+language output. For example
+
+```rust, ignore
+x.interpret("7+3*(2-1)", &output);
+```
+
+We don't claim that output assembly code is exactly what modern compilers generate,
+but it at least comuptes the expression correctly.
+It is simpy an example of using Interpreter pattern.
+
+## Solution 1
+
+The grammar for a set of espressions over `0,...,9, +,-,*,/,(,)` is
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> term * factor
+term -> term / factor
+term -> factor
+factor -> ( exp )
+factor -> 0 | 1| 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+Our first approach is a standard one, simple implementation of
+a recursive descent parser.The following code
```

Comment:
```suggestion
a recursive descent parser. The following code
```
First time I heard to use recursive descent for parsing, usually I heard it is used for machine learning.